controllers/clustercurator_controller.go

```diff
@@ -150,6 +150,9 @@ func newClusterCuratorPredicate() predicate.Predicate {
 				if newClusterCurator.Spec.CuratingJob != oldClusterCurator.Spec.CuratingJob && newClusterCurator.Spec.CuratingJob == "" {
 					return false
 				}
+				if oldClusterCurator.Spec.Upgrade.IntermediateUpdate != "" && oldClusterCurator.Spec.Upgrade.DesiredUpdate != "" {
```

Comment:
However, this is probably not enough to make the final update version "DesiredUpdate" immutable. Just because the reconciliation is not taking place doesn't mean the ClusterCurator CR is not potentially updated with a new value. When the non controller k8s jobs look at the ClusterCurator CR, they might see a new final update version "DesiredUpdate" that is different than the old one. 

One possible solution is stamp the `DesiredUpdate` into the Status so the k8s jobs are looking at the status for the locked DesiredUpdate value instead of spec. This will make it much harder for a user to make mistake and update the spec DesiredUpdate value. The status value can also have a message field to inform the user for EUS upgrade the intermediate and final versions are set in stone and they will need to create a new ClusterCurator to retry.

We can discuss async for options.

Reply:
Could you try with CEL again? Could you see below and see/test if all the cases are covered?

```
x-kubernetes-validations:
  - rule: "!(has(oldSelf.spec.immutableField) && self.spec.immutableField != oldSelf.spec.immutableField)"
    message: "The immutableField cannot be modified"
  - rule: "!(has(oldSelf.spec.immutableField) && oldSelf.spec.immutableField != '' && has(oldSelf.spec.conditionallyImmutableField) && self.spec.conditionallyImmutableField != oldSelf.spec.conditionallyImmutableField)"
    message: "The conditionallyImmutableField cannot be modified when immutableField exists"
  - rule: "!(has(self.spec.conditionallyImmutableField) && !has(oldSelf.spec.immutableField) && has(self.spec.immutableField))"
    message: "The immutableField cannot be added when conditionallyImmutableField exists"
  - rule: "!has(self.spec.immutableField) || has(self.spec.conditionallyImmutableField)"
    message: "The immutableField cannot exist if conditionallyImmutableField is missing"
```

- We want intermediateVersion to be immutable but if it was not present then it shouldn't be allowed to be added afterwards.
- We want finalVersion to be only immutable when intermediateVersion is present because we don't want to change the current behaviour that finalVersion is mutable.
- We don't want intermediateVersion by itself when finalVersion is missing

This way we can lock up both intermediateVersion and finalVersion when they are used together and not changed the existing behaviour of finalVersion being mutable if it's by itself.

If you go with this route, I recommend some golang unit testings to make sure we cover all the cases. Thanks!

Reply:
Thanks @mikeshng ! Let me try this out.

Reply:
I removed the validation on the `intermediateUpdate` field itself and added your validation to the `upgrade` object. I think this works perfectly now :) Thanks @mikeshng !

Reply:
@mikeshng I added some unit tests for this but they are failing. I'm guessing because the unit test is importing the schema from github and not from my local code so I commented the tests out. I assume they should work once the schema change is delivered to the repo.

Reply:
You can try something like this:
https://github.com/open-cluster-management-io/multicloud-operators-subscription/blob/main/pkg/controller/subscription/subscription_controller_suite_test.go#L41-L49
which will load the local CRD into envtest which will allow you to test some scenarios. 

The unit tests doesn't need to interact with any controllers or clustercurator k8s jobs.
I think we just need to ensure the CEL validations are working as expected. So load the CRD and then perform some create and update CR tests and verify expected error or ok.

Some of the tests (if any missing feel free to add):
- Create CR with just interVer, expect err
- Create CR with just finalVer, expect ok; update finalVer value, expect ok; update this CR by adding interVer, expect err
- Create CR with both interVer and finalVer, expect ok; update interVer, expect error; update finalVer, expect err; update both expect err; delete interVer, expect error; delete finalVer, expect err; delete both expect err

Reply:
Thanks @mikeshng ! Let me try this out. It might take some time since the unit tests doesn't have the kube-apiserver binaries configured.

Reply:
@mikeshng I created the unit tests. They are passing in the `ci/prow/unit-tests` job but failing in Sonar. I'm guessing because there's no way to setup EnvTest in the Sonar unit test? Should we exclude the tests in the Sonar unit test?

Reply:
I am not sure. Maybe we can ask @xiangjingli because we have similar tests in our appsub repo and they are fine in sonar.

Reply:
@fxiang1  I noticed that for each failed unit test function. the `m.Stop()` is called twice.  I guess the `defer m.Stop()` may fail in sonar as `m.Stop()` has been called. You may remove one of the Stop to see if it helps.

Reply:
Thanks @xiangjingli Looks like that's not the issue. My tests depend on a specific version of Kubebuilder 1.25 or greater for the CRD x-kubernetes-validation to work. In the `ci/prow/unit-tests` job we can download our own Kubebuilder version but in Sonar it's possible they have an older Kubebuilder version.

Reply:
How about we install the 1.25 kubebuilder before running sonar scan in our prow configuration yaml, sth like this 
```
    - name: Install Kubebuilder
      run: |
        KUBEBUILDER_VERSION=1.25
        os=$(go env GOOS)
        arch=$(go env GOARCH)
        curl -L -o kubebuilder.tar.gz https://github.com/kubernetes-sigs/kubebuilder/releases/download/v${KUBEBUILDER_VERSION}/kubebuilder_${KUBEBUILDER_VERSION}_${os}_${arch}.tar.gz
        tar -zxvf kubebuilder.tar.gz
        sudo mv kubebuilder_${KUBEBUILDER_VERSION}_${os}_${arch} /usr/local/kubebuilder
        export PATH=$PATH:/usr/local/kubebuilder/bin

```

https://github.com/openshift/release/blob/3eef787ecdcfdb32bab2607e6645a22cfef86bd8/ci-operator/config/stolostron/cluster-curator-controller/stolostron-cluster-curator-controller-main.yaml#L32

Reply:
Oh that's a good idea @xiangjingli . Let me try this. Thanks!